src/openms/source/ANALYSIS/OPENSWATH/PeakIntegrator.cpp

```diff
@@ -375,4 +375,706 @@ namespace OpenMS
+    // Return the maximum distance permitted for the Mean parameter, to avoid
+    // diverging from the optimal solution in gradient descent
+    return (max_pos - min_pos) * 0.35;
+  }
+
+  UInt PeakIntegrator::emg_gradient_descent(
+    const std::vector<double>& xs,
+    const std::vector<double>& ys,
+    double& best_h,
+    double& best_mu,
+    double& best_sigma,
+    double& best_tau
+  ) const
+  {
+    // Initial parameters
+    double h { *std::max_element(ys.begin(), ys.end()) };
+    double mu { computeInitialMean(xs, ys) };
+    double sigma { mu * 1e-2 };
+    double tau { sigma * 2.0 };
+
+    const UInt max_num_iter { 400000 }; // Maximum number of iterations
+    const double h_lower_boundary { h }; // Parameter `h` won't decrease below this value
+
+    std::vector<double> TrX, TrY; // Training set (positions and intensities)
+    if (!extractTrainingSet(xs, ys, TrX, TrY)) return 0;
+
+    // Variables containing the "previous" differentials
+    double prev_diff_E_h, prev_diff_E_mu, prev_diff_E_sigma, prev_diff_E_tau, previous_E;
+    prev_diff_E_h = prev_diff_E_mu = prev_diff_E_sigma = prev_diff_E_tau = 0.0;
+
+    // Part of computation in iRpropPlus()
+    // The parameter will change as much as these terms between iterations
+    double term_h, term_mu, term_sigma, term_tau;
+    term_h = term_mu = term_sigma = term_tau = 0.0;
+
+    // These variables will contain the values obtained at the best iteration
+    // The best iteration is decided by the smallest E found
+    // Therefore, here `best_E` is initialized with the maximum value for type `double`
+    double best_E;
+    previous_E = best_h = best_mu = best_sigma = best_tau = best_E = std::numeric_limits<double>::max();
+
+    // Keep track of the current iteration index, and the best iteration index
+    UInt iter_idx, best_iter;
+    iter_idx = best_iter = 0;
+
+    // This value will increase according to the number of iterations occurred,
+    // to avoid spamming the terminal with too much debug information
+    UInt info_iter_threshold { 1 };
+
+    // Learning rates (used in gradient descent and iRprop+)
+    double lr_h, lr_mu, lr_sigma, lr_tau;
+    lr_h = lr_mu = lr_sigma = lr_tau = 0.0125; // iRprop+ paper recommends 0.0125 // TODO: improve references to papers, maybe with something like [REF_NUM]
+
+    // Variables to limit the change in position `mu`
+    const double mu_max_dist = computeMuMaxDistance(TrX);
+    const double mu_left_boundary { mu - mu_max_dist };
+    const double mu_right_boundary { mu + mu_max_dist };
+
+    // The standard deviation between a selection of the precedent Es is computed.
+    // If said standard deviation is lower than a certain value,
+    // the computation of gradient descent is terminated
+    const Size last_few_Es_dim { 10 };
+    std::vector<double> last_few_Es(last_few_Es_dim, 0.0);
+    Size last_few_Es_idx = 0;
+    const double Es_std_dev_min = 1.0; // TODO: improve this value?
```

Comment:
I would have to understand the affects of different minimums to comment on this.  Perhaps testing different ranges in this value in the tests will provide further insight.